### ISSUE TITLE
Windows cannot resolve targetport name

### ISSUE THREAD

**Issue by ksubrmnn** [kubernetes/kubernetes#91783]
**What happened**: Windows nodes do not correctly resolve targetport when it is specified as a name

**What you expected to happen**: Window a kube proxy should select the correct port from the port list in the Service object

**How to reproduce it (as minimally and precisely as possible)**: deploy a ClusterIP service where the targetport is specified as the name of a port and not a port number. Loadbalancing will fail when the client is a Windows container regardless of where the service is deployed (Windows or Linux)

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): Applicable since Windows GA
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`): Windows 
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug): n/a
- Others:

**Comment by ksubrmnn** [comment]
@marosset FYI

**Comment by ksubrmnn** [comment]
@JocelynBerrendonner @daschott Can you please assign this issue to someone in core net? This impacted AKS customers. I can explain the fix, it is fairly simple

**Comment by daschott** [comment]
Sure

**Comment by wawa0210** [comment]
![image](https://user-images.githubusercontent.com/7907809/83933374-69535700-a7da-11ea-9d7a-a31a3a7e7614.png)

@ksubrmnn 

I am interested in this issue, can I follow up? I don't know if my steps to reproduce are correct. A deployment of a windows container is created, the containerPort is specified by the name `http`, and then this `http` is used for mapping in the service. Finally, access through nodeip:port. Everything looks normal

**Comment by wawa0210** [comment]
/assign

@ksubrmnn  @MikkelHegn 

I am very confused. The following are my latest and complete test steps, but they have not reproduced

An example of nginx, yaml is as follows:
```
apiVersion: v1
kind: Pod
metadata:
  name: nginx
  labels:
    component: nginx
spec:
  affinity:
    nodeAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        nodeSelectorTerms:
        - matchExpressions:
          - key: kubernetes.io/os
            operator: In
            values:
            - linux
  containers:
  - name: nginx
    image: nginx:1.19
    imagePullPolicy: IfNotPresent
    ports:
      - containerPort: 80
        name: http
        protocol: TCP
---
apiVersion: v1
kind: Service
metadata:
  labels:
    component: nginx
  name: nginx
spec:
  ports:
  - name: nginx-80
    port: 80
    protocol: TCP
    targetPort: http
  selector:
    component: nginx
  type: ClusterIP
```
It will create a pod running in linux and create a service at the same time, the results are as follows
```
[root@10-7-106-220 ~]# kubectl  apply -f nginx.yaml
pod/nginx created
service/nginx created
[root@10-7-106-220 ~]# kubectl get svc -o wide
NAME         TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE   SELECTOR
kubernetes   ClusterIP   10.96.0.1    <none>        443/TCP   8d    <none>
nginx        ClusterIP   10.96.2.54   <none>        80/TCP    6s    component=nginx
[root@10-7-106-220 ~]# kubectl get ep -o wide
NAME         ENDPOINTS            AGE
kubernetes   10.7.106.220:16443   8d
nginx        172.28.0.14:80       11s
[root@10-7-106-220 ~]# kubectl get po -o wide
NAME    READY   STATUS    RESTARTS   AGE     IP            NODE           NOMINATED NODE   READINESS GATES
nginx   1/1     Running   0          5m44s   172.28.0.14   10-7-106-220   <none>           <none>

1.  Test on Linux nodes in the cluster
```
[root@10-7-106-220 ~]# telnet 10.96.2.54 80
Trying 10.96.2.54...
Connected to 10.96.2.54.
Escape character is '^]'.
`` `
test passed

2. Test on Windows nodes in the cluster
```
PS C:\Users\Administrator> Test-NetConnection -Port 80 -ComputerName 10.96.2.54
ComputerName     : 10.96.2.54
RemoteAddress    : 10.96.2.54
RemotePort       : 80
InterfaceAlias   : vEthernet (Ethernet0)
SourceAddress    : 10.7.106.206
TcpTestSucceeded : True
```
test passed

3. Test on independent Windows container on Windows nodes  in the cluster
```
PS C:\Users\Administrator> docker exec -it 398b409d8570 powershell
Windows PowerShell
Copyright (C) Microsoft Corporation. All rights reserved.

PS C:\inetpub\wwwroot> Test-NetConnection -Port 80 -ComputerName 10.96.2.54
ComputerName     : 10.96.2.54
RemoteAddress    : 10.96.2.54
RemotePort       : 80
InterfaceAlias   : vEthernet (Ethernet)
SourceAddress    : 172.28.228.12
TcpTestSucceeded : True
```
ps. This is a separate container running via docker run based on asp.net mvc and test passed
----

So far I have not been able to reproduce this issue, have I forgotten any key steps??

I went to study the proxy code and found that the name type targetport has been processed

https://github.com/kubernetes/kubernetes/blob/35fc65dc2c614e57e5a019358a24af78ee1cfc16/pkg/proxy/winkernel/proxier.go#L1200-L1205

I am interested in this issue, can you help me reproduce it together

**Comment by MikkelHegn** [comment]
Original repro came from this Helm chart: https://github.com/helm/charts/tree/master/stable/redis using this values file: https://github.com/helm/charts/blob/master/stable/redis/values-production.yaml.

One difference I spot is that the `targetPort` on the service is the same as the name of the service (as well as the name of the port on the container).

**Comment by wawa0210** [comment]
@MikkelHegn 
I don’t seem to be able to reproduce the example provided by you. Can you reproduce the problem with a simpler nginx (I have nginx yaml in my last reply), after deployment, can you look at the corresponding svc entrypoint

**Comment by MikkelHegn** [comment]
@wawa0210 - I had an error in my repro, and I cannot repro this in my setup now. The simple deployment I did is similar to what you've tried above.

We need to go back and look at the original issue reported by one of our customers and see if ew can repro based on that. Stay tuned... :-)

@ksubrmnn - did you find a path in the code, which you suspected had some issues?

**Comment by ksubrmnn** [comment]
I didn't repro the issue myself so I've only looked at this secondhand, but I figured we could add a check here to see if the targetport is a string and then iterate through the ports to select the correct one. However, Windows only allows one port per endpoint so we should also be mindful of that.

If this issue is not reproducible, I can close it.

https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/winkernel/proxier.go#L261

**Comment by MikkelHegn** [comment]
Using this chart and values: https://github.com/helm/charts/tree/master/stable/redis using this values file: https://github.com/helm/charts/blob/master/stable/redis/values-production.yaml. I can reproduce not being able to reach the master service from a Windows node. It might be when all services run on a single Linux node, not 100% sure, need some more investigation to figure out what the problem is.

**Comment by wawa0210** [comment]
@ksubrmnn @MikkelHegn 

The test scenario I described earlier deploys nginx pods (only on linux nodes), and the windows node as a client is completely ok to communicate through clusterip. . I took the time today to try to deploy this redis and then test it

**Comment by ksubrmnn** [comment]
Hi, @wawa0210. This issue was a false repo. The issue was related to shared endpoints across multiple services. Significant refactor work was done in this area recently so we expect that the issue will be resolved in 1.19

**Comment by MikkelHegn** [comment]
This is the actual issue we encountered: https://github.com/kubernetes/kubernetes/issues/91703 - fix is already in...
